File: src/morphic/resizer/operations.py
"""
Image resize operations — fit, fill, stretch, and pad.

All operations preserve the original format by default and support
configurable output quality and format override.
"""

from __future__ import annotations

import logging
import os
from pathlib import Path

from PIL import Image, ImageOps

logger = logging.getLogger(__name__)

# Valid resize modes
RESIZE_MODES = ("fit", "fill", "stretch", "pad")


def resize_image(
    path: str,
    width: int,
    height: int,
    mode: str = "fit",
    output_folder: str | None = None,
    bg_color: str = "#000000",
    quality: int = 90,
    output_format: str | None = None,
) -> str:
    """Resize a single image.

    Parameters
    ----------
    path : str
        Path to the source image.
    width : int
        Target width in pixels.
    height : int
        Target height in pixels.
    mode : str
        Resize mode: ``"fit"`` (within bounds), ``"fill"`` (cover + crop),
        ``"stretch"`` (ignore ratio), ``"pad"`` (fit + pad borders).
    output_folder : str, optional
        Write output here instead of overwriting. Creates the folder
        if needed.
    bg_color : str
        Background colour for pad mode (CSS hex, default black).
    quality : int
        JPEG/WebP quality (1-100).
    output_format : str, optional
        Force output format (e.g. ``".png"``). Uses original if *None*.

    Returns
    -------
    str
        Path to the output file.

    Raises
    ------
    FileNotFoundError
        If the source file does not exist.
    ValueError
        If an invalid mode is given.
    """
    if not os.path.isfile(path):
        raise FileNotFoundError(f"File not found: {path}")
    if mode not in RESIZE_MODES:
        raise ValueError(
            f"Invalid mode '{mode}'. Must be one of {RESIZE_MODES}"
        )
    if width <= 0 or height <= 0:
        raise ValueError("Width and height must be positive integers")

    img = Image.open(path)

    # Convert palette/LA images to RGBA/RGB for processing
    if img.mode in ("P", "LA"):
        img = img.convert("RGBA")
    elif img.mode == "L":
        img = img.convert("RGB")

    size = (width, height)

    if mode == "fit":
        img.thumbnail(size, Image.Resampling.LANCZOS)
    elif mode == "fill":
        img = ImageOps.fit(img, size, Image.Resampling.LANCZOS)
    elif mode == "stretch":
        img = img.resize(size, Image.Resampling.LANCZOS)
    elif mode == "pad":
        img = ImageOps.pad(img, size, Image.Resampling.LANCZOS, color=bg_color)

    # Determine output path
    src = Path(path)
    ext = output_format if output_format else src.suffix
    if not ext.startswith("."):
        ext = f".{ext}"

    if output_folder:
        os.makedirs(output_folder, exist_ok=True)
        dest = Path(output_folder) / f"{src.stem}{ext}"
    else:
        dest = src.with_suffix(ext)

    # Convert RGBA to RGB for formats that don't support alpha
    if img.mode == "RGBA" and ext.lower() in (".jpg", ".jpeg", ".bmp"):
        img = img.convert("RGB")

    save_kwargs: dict = {}
    if ext.lower() in (".jpg", ".jpeg", ".webp"):
        save_kwargs["quality"] = quality
    if ext.lower() == ".png":
        save_kwargs["optimize"] = True

    img.save(str(dest), **save_kwargs)
    logger.info("Resized %s → %s (%s)", path, dest, mode)
    return str(dest)

File: internal/resizer/operations.go
package resizer

import (
	"fmt"
	"image"
	"image/color"
	"image/draw"
	"image/png"
	"os"
	"path/filepath"
	"strings"

	"github.com/disintegration/imaging"
)

// ResizeModes lists the valid resize mode strings.
var ResizeModes = []string{"fit", "fill", "stretch", "pad"}

// IsValidMode returns true if mode is one of the accepted values.
func IsValidMode(mode string) bool {
	for _, m := range ResizeModes {
		if m == mode {
			return true
		}
	}
	return false
}

// ResizeImage resizes a single image.
//
// Modes:
//   - "fit":     resize to fit within width×height, preserving aspect ratio
//   - "fill":    resize to cover width×height then center-crop
//   - "stretch": resize to exact width×height ignoring ratio
//   - "pad":     fit + pad borders with bgColor
//
// Returns the path to the output file.
func ResizeImage(
	path string,
	width, height int,
	mode string,
	outputFolder string,
	bgColor string,
	quality int,
	outputFormat string,
) (string, error) {
	if _, err := os.Stat(path); err != nil {
		return "", fmt.Errorf("file not found: %s", path)
	}
	if !IsValidMode(mode) {
		return "", fmt.Errorf("invalid mode '%s'", mode)
	}
	if width <= 0 || height <= 0 {
		return "", fmt.Errorf("width and height must be positive")
	}

	img, err := imaging.Open(path)
	if err != nil {
		return "", err
	}

	var result *image.NRGBA

	switch mode {
	case "fit":
		result = imaging.Fit(img, width, height, imaging.Lanczos)
	case "fill":
		result = imaging.Fill(img, width, height, imaging.Center, imaging.Lanczos)
	case "stretch":
		result = imaging.Resize(img, width, height, imaging.Lanczos)
	case "pad":
		fitted := imaging.Fit(img, width, height, imaging.Lanczos)
		bg := parseHexColor(bgColor)
		canvas := imaging.New(width, height, bg)
		offsetX := (width - fitted.Bounds().Dx()) / 2
		offsetY := (height - fitted.Bounds().Dy()) / 2
		draw.Draw(canvas, image.Rect(offsetX, offsetY,
			offsetX+fitted.Bounds().Dx(), offsetY+fitted.Bounds().Dy()),
			fitted, image.Point{}, draw.Over)
		result = canvas
	}

	// Determine output path
	srcExt := filepath.Ext(path)
	ext := outputFormat
	if ext == "" {
		ext = srcExt
	}
	if !strings.HasPrefix(ext, ".") {
		ext = "." + ext
	}

	stem := strings.TrimSuffix(filepath.Base(path), srcExt)
	var dest string
	if outputFolder != "" {
		os.MkdirAll(outputFolder, 0755)
		dest = filepath.Join(outputFolder, stem+ext)
	} else {
		dest = filepath.Join(filepath.Dir(path), stem+ext)
	}

	// imaging.Save handles format detection from extension
	opts := []imaging.EncodeOption{}
	extLower := strings.ToLower(ext)
	if extLower == ".jpg" || extLower == ".jpeg" || extLower == ".webp" {
		opts = append(opts, imaging.JPEGQuality(quality))
	}
	if extLower == ".png" {
		opts = append(opts, imaging.PNGCompressionLevel(png.DefaultCompression))
	}

	if err := imaging.Save(result, dest, opts...); err != nil {
		return "", err
	}

	return dest, nil
}

// parseHexColor parses a CSS hex colour like "#FF0000" to a color.NRGBA.
func parseHexColor(s string) color.NRGBA {
	s = strings.TrimPrefix(s, "#")
	if len(s) != 6 {
		return color.NRGBA{A: 255} // fallback to black
	}
	var r, g, b uint8
	fmt.Sscanf(s, "%02x%02x%02x", &r, &g, &b)
	return color.NRGBA{R: r, G: g, B: b, A: 255}
}
